README.md
# Mage2Docker

> Get rid of your "it works on my machine" arguments.

This project intends to create an opinionated, performant, replicable, and production-like Magento 2 local development environment that leverages [Docker](https://www.docker.com/), [VSCode](https://code.visualstudio.com/) and [devcontainers](https://code.visualstudio.com/docs/remote/containers).

<div align="center">

[![graycore/magento2-php pulls](https://img.shields.io/docker/pulls/graycore/magento-php.svg?label=magento-php%20docker%20pulls)](https://hub.docker.com/r/graycore/magento-php)
[![Mage2Docker Commit Activity](https://img.shields.io/badge/maintained%3F-yes-brightgreen.svg)](https://github.com/graycoreio/mage2docker/graphs/commit-activity)
[![Mage2Docker MIT License](https://img.shields.io/badge/license-MIT-blue.svg)](https://github.com/graycoreio/mage2docker/blob/master/LICENSE.md)
[![Mappia Discord](https://img.shields.io/discord/908816550121857074?color=11689b)](http://chat.mappia.io/)

</div>

### Dolphin
As of Mage2Docker v6.0.0, Dolphin is the new fully cross-platform Mage2docker environment that leverages VSCode `devcontainers` or Github Codespaces.

To get started, see [the new significantly smoother installation process. We strongly encourage you to use it](./docs/stories/dolphin.md).

For those users curious on why we've made these changes, [please read the architecture discussion.](./docs/arch/dolphin.md)

## Prerequisites

#### Hardware
* 4GB RAM
* 2 Cores

## Features

* :clock1: [**5 Minute** Setup](#supported-platforms)
  * :octocat: Bring your own repo (BYOR)
  * :biohazard:	Use Magento Source Code
  * :violin: Using [Composer](https://getcomposer.org/)
* :fire: Fast (~250ms Response Times Out-of-the-box)
* :evergreen_tree: Alpine Linux
* :grin: [Semver Compliant](https://semver.org/)
* Magento 2 Open Source & Commerce (Not Cloud) Support
  * If you're using cloud, please use the ["Cloud Docker" project created by Adobe](https://devdocs.magento.com/cloud/docker/docker-config.html)
* :lock: [Local TLS Certificates](./docs/stories/ssl/making-tls-work-locally.md)
* :gear: [Infinitely Configurable](./docs/stories/configuring.md)
* [VSCode](https://code.visualstudio.com/)
* [Codespaces ready](https://github.com/features/codespaces)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [6.0.0-alpha.4](https://github.com/graycoreio/mage2docker/compare/v6.0.0-alpha.3...v6.0.0-alpha.4) (2022-08-17)


### Features

* add patch to base cli and fpm images ([#78](https://github.com/graycoreio/mage2docker/issues/78)) ([a77b64a](https://github.com/graycoreio/mage2docker/commit/a77b64a981bb7f9edef1d02ef01fb554a50d4d2f))

## [6.0.0-alpha.3](https://github.com/graycoreio/mage2docker/compare/v5.0.0...v6.0.0-alpha.3) (2022-08-07)


### Features

* add a devcontainer.json that uses dolphin ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([708fbe7](https://github.com/graycoreio/mage2docker/commit/708fbe715d9c661ceab43d81a4f88d062728be99))
* add images for php 8.1 ([#70](https://github.com/graycoreio/mage2docker/issues/70)) ([8d25097](https://github.com/graycoreio/mage2docker/commit/8d250976755099467e9d1b6fa9afd5dff85ba388))
* adding elasticsearch 7.16.1 ([#51](https://github.com/graycoreio/mage2docker/issues/51)) ([5127f9d](https://github.com/graycoreio/mage2docker/commit/5127f9d982884d46b2228f0c0cd85a2d5015a20c))
* enforce `yml` file names ([#75](https://github.com/graycoreio/mage2docker/issues/75)) ([4e2d333](https://github.com/graycoreio/mage2docker/commit/4e2d33313deabe361435979d9f76ba25ac49fa98))
* **env:** add revised ENV for dolphin ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([9020d1e](https://github.com/graycoreio/mage2docker/commit/9020d1eca8856fac1052791810ca5b6609c6c801))
* **nginx:** add auto-reloading ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([8637597](https://github.com/graycoreio/mage2docker/commit/863759737bdca5a96c0a753ddcd363d9cc8773a9))
* **php:** add new dolphin base images for php 7.2, 7.3, and 7.4 ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([db7a950](https://github.com/graycoreio/mage2docker/commit/db7a9501f448b9b38a344aa30418c4b63c1624de))
* **php:** add, use new configurations for dolphin php ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([d4d1c78](https://github.com/graycoreio/mage2docker/commit/d4d1c7892136c65e6a322a3839e5231fa34064f2))
* **sync:** add new container sync strategy ([#64](https://github.com/graycoreio/mage2docker/issues/64)) ([35fe4db](https://github.com/graycoreio/mage2docker/commit/35fe4dbc783e400638c204e6dba2b0e470933ade))


### Bug Fixes

* dont restart mysql always, only on failure ([#44](https://github.com/graycoreio/mage2docker/issues/44)) ([c999ac2](https://github.com/graycoreio/mage2docker/commit/c999ac265b10c5db3e69370206bccc040e6571f9))
* limit elasticsearch memory consumption ([#54](https://github.com/graycoreio/mage2docker/issues/54)) ([5521a05](https://github.com/graycoreio/mage2docker/commit/5521a057dbcd06c46c3620210d431e526e964de6))

## [5.0.0](https://github.com/graycoreio/mage2docker/compare/v4.1.0...v5.0.0) (2021-09-27)


### ⚠ BREAKING CHANGES

* **php:** composer 2 is now installed in the php containers
* **configuration:** The default COMPOSE_FILE path has changed and the new root docker-compose.yml is empty. To fix your environment, see the updated env.sample. Likely, all you will have to do is append "./compose/base/2.3.yml" to your existing COMPOSE_FILE. If you run into problems, please report a Github issue.

### Features

* **configuration:** allow swapping between 2.3 and 2.4 more easily. ([3c5ada3](https://github.com/graycoreio/mage2docker/commit/3c5ada3309044d8e0e8329ff10f26ccee2274431))
* **php:** 7.4, 7.3 now use composer 2 ([#32](https://github.com/graycoreio/mage2docker/issues/32)) ([3fc5c95](https://github.com/graycoreio/mage2docker/commit/3fc5c9581ef7c2bf3184525a88ebb14b45971bc7))


### Bug Fixes

* **php, 7.4:** add a missing hyphen in a docker image name in compose file ([13dcdc1](https://github.com/graycoreio/mage2docker/commit/13dcdc128c950e3c409442280b514efd316ce2be))

## [4.1.0](https://github.com/graycoreio/mage2docker/compare/v4.0.1...v4.1.0) (2021-05-11)


### Features

* **mysql:** ignore database files in entrypoint to prevent overcomitting ([#21](https://github.com/graycoreio/mage2docker/issues/21)) ([bc1fae2](https://github.com/graycoreio/mage2docker/commit/bc1fae21586d1a92afd335ccd46e99be581622d2))


### Bug Fixes

* **nginx, 1.18:** add missing default env variables ([ac225ce](https://github.com/graycoreio/mage2docker/commit/ac225cee4dce6134296f88f24c8ea2355783d1d2))

### [4.0.1](https://github.com/graycoreio/mage2docker/compare/v4.0.0...v4.0.1) (2021-02-10)

## [4.0.0](https://github.com/graycoreio/mage2docker/compare/v3.5.0...v4.0.0) (2021-02-10)


### ⚠ BREAKING CHANGES

* **nginx:** drop support for v1.17

### Features

* **nginx:** add support for nginx v1.19 ([#18](https://github.com/graycoreio/mage2docker/issues/18)) ([3ec7152](https://github.com/graycoreio/mage2docker/commit/3ec71522ec1a34abb6f5040fd18965a43974835b))

## [3.5.0](https://github.com/graycoreio/mage2docker/compare/v3.4.0...v3.5.0) (2020-12-15)


### Features

* **magento2:** add envfile to ENV of magento container ([6ad018d](https://github.com/graycoreio/mage2docker/commit/6ad018de9ad95ebcf7defc4ef5e9dd505a6e271a))


### Bug Fixes

* **php,7.2,7.3,7.4:** pin composer to version 1 ([81b1a61](https://github.com/graycoreio/mage2docker/commit/81b1a61acec7de24e50be47e5349f55a9094f345))

## [3.4.0](https://github.com/graycoreio/mage2docker/compare/v3.3.0...v3.4.0) (2020-11-24)


### Features

* **php:** make php.ini settings more production like by default ([8d982be](https://github.com/graycoreio/mage2docker/commit/8d982be164f1ce9aea400920d55cbf9dea1753a8))

## [3.3.0](https://github.com/graycoreio/mage2docker/compare/v3.2.0...v3.3.0) (2020-11-06)


### Features

* **docker-develop:** adding mailhog service to docker-compose ([36a649d](https://github.com/graycoreio/mage2docker/commit/36a649dadd546397214b92db1fb0093efeeceb25))

tools/standard-version/compose-yml.js
const yaml = require('yaml');

const semverMatchRegex = /-v(\d*.*)$/;

module.exports.readVersion = function (contents) {
  const data = yaml.parse(contents);

  for(let key of Object.keys(data.services)) {
    if(data.services[key]?.image?.includes("graycore")){
      const matchedVersion = data.services[key]?.image.match(semverMatchRegex)[1];
      if(matchedVersion){
        return matchedVersion;
      }
    }
  }

  throw new Error("Unable to find version");
}

module.exports.writeVersion = function (contents, version) {
  const data = yaml.parse(contents);

  for(let key of Object.keys(data.services)) {
    if(data.services[key]?.image?.includes("graycore")){
      const matchedVersion = data.services[key]?.image.match(semverMatchRegex)[1];
      console.lo
      data.services[key].image = data.services[key].image.replace(matchedVersion, version);
    }
  }

  return yaml.stringify(data);
}


.github/workflows/startup/configure.sh
#!/bin/bash
set -e

MAGENTO_VERSION_2_3='magento/project-community-edition=2.3.*'
MAGENTO_VERSION_2_4='magento/project-community-edition=2.4.*'
COMPOSE_PROJECT_NAME='mage2docker';

case "$COMPOSER_PROJECT" in
   "$MAGENTO_VERSION_2_3")
        printf "\\n\\nCOMPOSE_FILE=docker-compose.yml:./compose/base/2.3.yml:compose/sync/containersync.yml:compose/php/7.4/build.yml" >> .env
        sed -i -e "s|^[#]*\s*COMPOSER_PROJECT=.*|COMPOSER_PROJECT="$COMPOSER_PROJECT"|" .env
      ;;
   $MAGENTO_VERSION_2_4)
        printf "\\n\\nCOMPOSE_FILE=docker-compose.yml:./compose/base/2.4.yml:compose/sync/containersync.yml:compose/php/7.4/build.yml" >> .env
        sed -i -e "s|^[#]*\s*COMPOSER_PROJECT=.*|COMPOSER_PROJECT="$COMPOSER_PROJECT"|" .env
      ;;
   *)
        printf "Unknown Magento version"
        exit 1;
     ;;
esac

sed -i -e "s|^[#]*\s*COMPOSE_PROJECT_NAME=.*|COMPOSE_PROJECT_NAME="$COMPOSE_PROJECT_NAME"|" .env

env.docker.php
<?php
return [
    'backend' => [
        'frontName' => 'admin'
    ],
    'crypt' => [
        'key' => 'REPLACEME'
    ],
    'db' => [
        'table_prefix' => '',
        'connection' => [
            'default' => [
                'host' => 'database',
                'dbname' => 'magento2',
                'username' => 'magento2',
                'password' => 'magento2',
                'active' => '1',
                'model' => 'mysql4',
                'engine' => 'innodb',
                'initStatements' => 'SET NAMES utf8;'
            ]
        ]
    ],
    'resource' => [
        'default_setup' => [
            'connection' => 'default'
        ]
    ],
    'x-frame-options' => 'SAMEORIGIN',
    'MAGE_MODE' => 'developer',
    'session' => [
        'save' => 'redis',
        'redis' => [
            'host' => 'sessioncache',
            'port' => '6379',
            'password' => '',
            'timeout' => '2.5',
            'persistent_identifier' => '',
            'database' => '0',
            'compression_threshold' => '2048',
            'compression_library' => 'gzip',
            'log_level' => '4',
            'max_concurrency' => '6',
            'break_after_frontend' => '5',
            'break_after_adminhtml' => '30',
            'first_lifetime' => '600',
            'bot_first_lifetime' => '60',
            'bot_lifetime' => '7200',
            'disable_locking' => '0',
            'min_lifetime' => '60',
            'max_lifetime' => '2592000',
            'sentinel_master' => '',
            'sentinel_servers' => '',
            'sentinel_connect_retries' => '5',
            'sentinel_verify_master' => '0'
        ]
    ],
    'cache' => [
        'frontend' => [
            'default' => [
                'id_prefix' => '40d_',
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => 'cache',
                    'database' => '0',
                    'port' => '6379',
                    'password' => '',
                    'compress_data' => '1',
                    'compression_lib' => ''
                ]
            ],
            'page_cache' => [
                'id_prefix' => '40d_',
                'backend' => 'Cm_Cache_Backend_Redis',
                'backend_options' => [
                    'server' => 'fullpagecache',
                    'database' => '0',
                    'port' => '6379',
                    'password' => '',
                    'compress_data' => '0',
                    'compression_lib' => ''
                ]
            ]
        ]
    ],
    'lock' => [
        'provider' => 'db',
        'config' => [
            'prefix' => ''
        ]
    ],
    'cache_types' => [
        'config' => 1,
        'layout' => 1,
        'block_html' => 1,
        'collections' => 1,
        'reflection' => 1,
        'db_ddl' => 1,
        'compiled_config' => 1,
        'eav' => 1,
        'customer_notification' => 1,
        'config_integration' => 1,
        'config_integration_api' => 1,
        'full_page' => 1,
        'config_webservice' => 1,
        'translate' => 1,
        'vertex' => 1
    ],
    'install' => [
        'date' => 'Mon, 01 Jul 2019 18:06:52 +0000'
    ]
];
